billboards using geometry shader

m-agour opened the pull request:
This is an example on how a geometry shader can be used along with some minimal calculations to generate billboards using the `dot` actor.

https://user-images.githubusercontent.com/63170874/179493243-55b28d24-2c94-485d-af7e-ccb296733f34.mp4

m-agour committed:
Week 3 blog post.

m-agour committed:
Changed video size

m-agour committed:
Fixed PR number

ganimtron-10 committed:
adding template

ganimtron-10 committed:
adding content

xtanion committed:
Adding week 3 blog

xtanion committed:
added images

ganimtron-10 committed:
adding links to the changes

m-agour committed:
Fixed typo, width of the video and some marking issues

xtanion committed:
typos fixes and renaming previous blog

ganimtron-10 committed:
adding punctuation

m-agour committed:
Fixed some issues.

skoudoro committed:
skip test_frame_rate on ubuntu

xtanion committed:
doc fixes

m-agour committed:
Geometry shader to draw quads (for billboards later)

m-agour committed:
Geometry shader to draw quads (for billboards later)

m-agour committed:
example to use geom_quad

m-agour committed:
example to use geom_quad

m-agour committed:
geom_quad

m-agour committed:
geom_quad

m-agour committed:
fixed scale

m-agour committed:
Added billboard_gs actor

m-agour committed:
Simplified the example

m-agour committed:
Merge branch 'Using-gemoetry-shader-to-generate-primitives' of https://github.com/m-agour/fury into Using-gemoetry-shader-to-generate-primitives

m-agour committed:
billboard geometry using geometry shader

guaje reviewed (changes requested):
@m-agour, thanks for exploring this.

This feature was proposed as part of an entire [GSoC project](https://github.com/fury-gl/fury/wiki/Google-Summer-of-Code-2022-(GSOC2022)#project-3-tessellated-like-actors). Since this feature requires many things taken into account, we'll use this PR to ensure the bare minimum compared to the current state of the billboard actor, so we don't affect the tasks of your main GSoC project.

I'll add @tvcastillod to this conversation since this PR is related to her project.

guaje commented on fury/actor.py:
```diff
@@ -2383,6 +2383,17 @@ def billboard(centers, colors=(0, 1, 0), scales=1, vs_dec=None, vs_impl=None,
     return bb_actor
 
 
+def billboard_gs(centers, colors, scales=(1, 1, 1)):
```
@filipinascimento I think it would be better to have both functionalities. We have seen how some things break when there's a major release of VTK, especially things related to shaders. If that happens in the future, we'll just change the default parameter to the original functionality while we fix the shaders.

Regarding the properties, the purpose of changing the dot actor in #584 was precisely this, using a new dot actor as the basis for a GS version of the billboard. In that case, the colors are handled by the polydata in the dot actor.

m-agour committed:
Moved billboard actor

m-agour committed:
renamed gs shader file

m-agour committed:
Added a small example

m-agour committed:
GS-based-billboard shaders

m-agour committed:
sphere frag impl shader

m-agour committed:
removed old tutorial

m-agour committed:
Merge branch 'Using-gemoetry-shader-to-generate-primitives' into Billboards-using-geometry-shader

m-agour committed:
removed old actor function

m-agour committed:
updated tutorials

m-agour committed:
unwanted shader

pep8speaks commented:
Hello @m-agour! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`docs/tutorials/03_shaders/viz_billboard_sdf_spheres.py`](https://github.com/fury-gl/fury/blob/f0029335d682252882b1c774d63ec96280563dc2/docs/tutorials/03_shaders/viz_billboard_sdf_spheres.py):

> [Line 61:1](https://github.com/fury-gl/fury/blob/f0029335d682252882b1c774d63ec96280563dc2/docs/tutorials/03_shaders/viz_billboard_sdf_spheres.py#L61): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 64:1](https://github.com/fury-gl/fury/blob/f0029335d682252882b1c774d63ec96280563dc2/docs/tutorials/03_shaders/viz_billboard_sdf_spheres.py#L64): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 70:1](https://github.com/fury-gl/fury/blob/f0029335d682252882b1c774d63ec96280563dc2/docs/tutorials/03_shaders/viz_billboard_sdf_spheres.py#L70): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 73:1](https://github.com/fury-gl/fury/blob/f0029335d682252882b1c774d63ec96280563dc2/docs/tutorials/03_shaders/viz_billboard_sdf_spheres.py#L73): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 82:1](https://github.com/fury-gl/fury/blob/f0029335d682252882b1c774d63ec96280563dc2/docs/tutorials/03_shaders/viz_billboard_sdf_spheres.py#L82): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 85:1](https://github.com/fury-gl/fury/blob/f0029335d682252882b1c774d63ec96280563dc2/docs/tutorials/03_shaders/viz_billboard_sdf_spheres.py#L85): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 94:1](https://github.com/fury-gl/fury/blob/f0029335d682252882b1c774d63ec96280563dc2/docs/tutorials/03_shaders/viz_billboard_sdf_spheres.py#L94): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 97:1](https://github.com/fury-gl/fury/blob/f0029335d682252882b1c774d63ec96280563dc2/docs/tutorials/03_shaders/viz_billboard_sdf_spheres.py#L97): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 113:1](https://github.com/fury-gl/fury/blob/f0029335d682252882b1c774d63ec96280563dc2/docs/tutorials/03_shaders/viz_billboard_sdf_spheres.py#L113): [E122](https://duckduckgo.com/?q=pep8%20E122) continuation line missing indentation or outdented
> [Line 132:1](https://github.com/fury-gl/fury/blob/f0029335d682252882b1c774d63ec96280563dc2/docs/tutorials/03_shaders/viz_billboard_sdf_spheres.py#L132): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 135:1](https://github.com/fury-gl/fury/blob/f0029335d682252882b1c774d63ec96280563dc2/docs/tutorials/03_shaders/viz_billboard_sdf_spheres.py#L135): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 146:1](https://github.com/fury-gl/fury/blob/f0029335d682252882b1c774d63ec96280563dc2/docs/tutorials/03_shaders/viz_billboard_sdf_spheres.py#L146): [E122](https://duckduckgo.com/?q=pep8%20E122) continuation line missing indentation or outdented
> [Line 163:1](https://github.com/fury-gl/fury/blob/f0029335d682252882b1c774d63ec96280563dc2/docs/tutorials/03_shaders/viz_billboard_sdf_spheres.py#L163): [E122](https://duckduckgo.com/?q=pep8%20E122) continuation line missing indentation or outdented
> [Line 193:1](https://github.com/fury-gl/fury/blob/f0029335d682252882b1c774d63ec96280563dc2/docs/tutorials/03_shaders/viz_billboard_sdf_spheres.py#L193): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '

m-agour committed:
Additional files came with merging from Javier's PR

m-agour commented:
Hello @guaje @filipinascimento ,
I updated this PR and changed the tutorial to display impostor sphere actors on billboards as shown below

https://user-images.githubusercontent.com/63170874/182064895-27fdd00a-6372-4caa-aff6-3a4bad64e407.mp4

I will make an example of animating (scaling and changing the color of spheres) before today's meeting. Also will put it under the pressure of thousands of spheres at once.

m-agour committed:
Added example for testing animation on bbgs

m-agour committed:
Added intersection (depth) to spheres

guaje commented on fury/actor.py:
```diff
@@ -2350,6 +2352,23 @@ def billboard(centers, colors=(0, 1, 0), scales=1, vs_dec=None, vs_impl=None,
+                        block='prim_id')
+
+        shader_to_actor(bb_actor, 'fragment',
+                        decl_code=import_fury_shader('gs_billboard_dec.frag'),
```
Add support to the custom shaders in the params `fs_dec` and `fs_impl`. As implemented [here](https://github.com/fury-gl/fury/blob/2b2caf849efd2fa894d33be661a8dd3d6679c6b2/fury/actor.py#L2371) and [here](https://github.com/fury-gl/fury/blob/2b2caf849efd2fa894d33be661a8dd3d6679c6b2/fury/actor.py#L2373).

guaje commented on fury/actor.py:
```diff
@@ -2350,6 +2352,23 @@ def billboard(centers, colors=(0, 1, 0), scales=1, vs_dec=None, vs_impl=None,
+                        block='prim_id')
+
+        attribute_to_actor(bb_actor, scales.flatten(), 'scale')
+        bb_actor.GetProperty().BackfaceCullingOff()
+        bb_actor.GetMapper().SetVBOShiftScaleMethod(False)
```
These functions are used in both billboard creation approaches.

guaje commented on fury/actor.py:
```diff
@@ -2317,8 +2317,8 @@ def have_2_dimensions(arr):
 
-def billboard(centers, colors=(0, 1, 0), scales=1, vs_dec=None, vs_impl=None,
-              gs_prog=None, fs_dec=None, fs_impl=None):
+def billboard(centers, colors=(0, 1, 0), scales=1, using_gs=False, vs_dec=None,
```
Maybe `use_gs` instead of `using_gs`.

guaje commented on fury/actor.py:
```diff
@@ -2317,8 +2317,8 @@ def have_2_dimensions(arr):
-def billboard(centers, colors=(0, 1, 0), scales=1, vs_dec=None, vs_impl=None,
-              gs_prog=None, fs_dec=None, fs_impl=None):
+def billboard(centers, colors=(0, 1, 0), scales=1, using_gs=False, vs_dec=None,
+              vs_impl=None, gs_prog=None, fs_dec=None, fs_impl=None):
```
Given the nature of the new functionality, we should find a way to reuse the `gs_prog` param. Something like:

```python
def billboard(..., gs_prog='gs_generated', ...):
    if gs_prog.lower() == 'gs_generated':
        # Your super awesome GS Billboard implementation
    elif gs_prog is None:
        # Use the current billboard functionality
    else:
        # Use the user's custom GS program
```

This way, we can dispose of the `use_gs/using_gs` parameter.

guaje reviewed (changes requested):
@m-agour these changes look much better. See my suggestions on better API integration. We still need to add a proper test and make sure we're scaling the fragments to keep the compatibility with the current implementation.

m-agour committed:
Revert "Additional files came with merging from Javier's PR"

This reverts commit 52d42103458dd7ebb1f5f1bcfe7061963edbb4da.

m-agour committed:
Moved to separate file

m-agour committed:
r is separated from center

m-agour committed:
composed fs dec

m-agour committed:
renamed scale uniform

m-agour committed:
wrapped in else

m-agour committed:
Renamed atrribs

m-agour committed:
updated lightning calculations

m-agour committed:
missing rename

m-agour committed:
Allowed adding custom shader code

m-agour committed:
reused gs_prog param

m-agour committed:
renamed gs_prog case to default

m-agour committed:
Translated spheres

m-agour committed:
unsuccessful antialiasing

Alpha blending does not work correctly

m-agour committed:
Update uniforms values

m-agour committed:
removed additional shader file

m-agour committed:
updated calculations

m-agour committed:
Added lightColor0 uniform setter

still needs some work

skoudoro commented:
Can you rebase and fix the conflicts of this PR? 

Thank you

m-agour committed:
Merge branch 'master' into Billboards-using-geometry-shader

m-agour committed:
removed returns in shaders as @guaje requested

m-agour committed:
Fixed lights setting

m-agour committed:
Unwanted print

m-agour committed:
repeated lines of code

m-agour committed:
`gs_default` and get scene different way

m-agour committed:
`gs_prog` Back to Noe

m-agour committed:
scale consistency

m-agour committed:
Updated tests
